src/graph/sampler.cc

```diff
@@ -1592,17 +1629,31 @@ class WeightedEdgeSamplerObject: public EdgeSamplerObject {
   }
 
   void Fetch(DGLRetValue* rv) {
+    const int64_t num_workers = std::min(num_workers_, max_batch_id_ - curr_batch_id_);
     // generate subgraphs.
-    std::vector<SubgraphRef> positive_subgs(num_workers_);
-    std::vector<SubgraphRef> negative_subgs(num_workers_);
+    std::vector<SubgraphRef> positive_subgs(num_workers);
+    std::vector<SubgraphRef> negative_subgs(num_workers);
+
 #pragma omp parallel for
-    for (int i = 0; i < num_workers_; i++) {
+    for (int i = 0; i < num_workers; i++) {
       const dgl_id_t *seed_edge_ids = static_cast<const dgl_id_t *>(seed_edges_->data);
-      std::vector<int64_t> edge_ids(batch_size_);
+      std::vector<size_t> edge_ids(batch_size_);
 
-      for (int i = 0; i < batch_size_; ++i) {
-        int64_t edge_id = edge_selector_->Sample();
-        edge_ids[i] = seed_edge_ids[edge_id];
+#pragma omp critical
+      {
+        if (replacement_ == false) {
+          size_t n = batch_size_;
+          size_t num_ids = edge_selector_->SampleWithoutReplacement(n, &edge_ids);
+          while (edge_ids.size() > num_ids) {
+            edge_ids.pop_back();
+          }
```

Comment:
I'm not sure what it means. Why do you need to pop edge Ids?

Reply:
SampleWithoutReplacement may samples items less than required in “n”. （WithoutReplacement will not return sampled items, so at the last minibatch, it may return items < batchsize)